Sources/SwiftDocC/Utility/Graphs/DirectedGraph+Traversal.swift

```diff
@@ -0,0 +1,105 @@
+    func depthFirstSearch(from startingPoint: Node) -> some Sequence<Node> {
+        IteratorSequence(GraphNodeIterator(from: startingPoint, traversal: .depthFirst, in: self))
+    }
+}
+
+extension DirectedGraph {
+    /// A path through the graph, including the start and end nodes.
+    typealias Path = [Node]
+    /// Information about the current accumulated path during iteration.
+    typealias PathIterationElement = (path: Path, isLeaf: Bool, cycleStartIndex: Int?)
+    
+    /// Returns a sequence of accumulated path information from traversing the graph in breadth first order.
+    func accumulatingPaths(from startingPoint: Node) -> some Sequence<PathIterationElement> {
+        IteratorSequence(GraphBreadthFirstPathIterator(from: startingPoint, in: self))
+    }
+}
+
+// MARK: Node iterator
+
+/// An iterator that traverses a graph in either breadth first or depth first order depending on the buffer it uses to track nodes to traverse next.
+private struct GraphNodeIterator<Node: Hashable>: IteratorProtocol {
+    enum Traversal {
+        case breadthFirst, depthFirst
+    }
+    var traversal: Traversal
+    var graph: DirectedGraph<Node>
+    
+    private var nodesToTraverse: [Node]
+    private var seen: Set<Node>
+    
+    init(from startingPoint: Node, traversal: Traversal, in graph: DirectedGraph<Node>) {
+        self.traversal = traversal
+        self.graph = graph
+        self.nodesToTraverse = [startingPoint]
+        self.seen = []
+    }
+    
+    private mutating func pop() -> Node? {
+        guard !nodesToTraverse.isEmpty else { return nil }
+        
+        switch traversal {
+        case .breadthFirst:
+            return nodesToTraverse.removeFirst()
+        case .depthFirst:
+            return nodesToTraverse.removeLast()
+        }
+    }
+    
+    mutating func next() -> Node? {
+        while let node = pop() {
+            guard !seen.contains(node) else { continue }
+            seen.insert(node)
+            
+            nodesToTraverse.append(contentsOf: graph.neighbors(of: node))
+            
+            return node
+        }
+        return nil
+    }
+}
+
+// MARK: Path iterator
+
+/// An iterator that traverses a graph in breadth first order and returns information about the accumulated path through the graph, up to the current node.
+private struct GraphBreadthFirstPathIterator<Node: Hashable>: IteratorProtocol {
+    var pathsToTraverse: [(Node, [Node])]
+    var graph: DirectedGraph<Node>
+    
+    init(from startingPoint: Node, in graph: DirectedGraph<Node>) {
+        self.pathsToTraverse = [(startingPoint, [])]
+        self.graph = graph
+    }
+    
+    mutating func next() -> DirectedGraph<Node>.PathIterationElement? {
+        guard !pathsToTraverse.isEmpty else { return nil }
+        let (node, path) = pathsToTraverse.removeFirst()
+        
+        if let cycleStartIndex = path.lastIndex(of: node) {
```

Comment:
> I'm guessing this is an important detail.

This actually doesn't matter for correctness at all :) 

I'm simply using `lastIndex(of:)` instead of `firstIndex(of:)` to make the search from the end instead of from the start. The idea was that the cycle is probably small, meaning that the cycle point would be closer to the end than to the start. However, now that I write this, I realize that it's optimizing for the wrong thing since cycles are the rare case and there's going to be many more cases where the algorithm needs to traverse the full path to determine that there _isn't_ a cycle. I'll change it to `firstIndex(of:)` instead.